crazy_rl/multi_agent/jax/base_parallel_env.py

```diff
@@ -0,0 +1,151 @@
+        """Returns the observation space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _action_space(self, agent: int) -> spaces.Space:
+        """Returns the action space of the environment. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_obs(self, state: State) -> State:
+        """Computes the current observation of the environment from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _transition_state(self, state: State, actions: jnp.ndarray, key: jnp.ndarray) -> State:
+        """Transitions the state based on the mechanics of the environment, for example makes the target move and sanitize the actions. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _sanitize_action(self, state: State, actions: jnp.ndarray) -> State:
+        """Makes the actions passed to step fit the environment, e.g. avoid making brutal moves. Must be implemented in a subclass.
+
+        Args:
+            state : the state of the environment (contains agent_location used in this function).
+            actions : 2D array containing the x, y, z action for each drone.
+        """
+        raise NotImplementedError
+
+    def _compute_reward(self, state: State) -> State:
+        """Computes the current reward value(s) from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_terminated(self, state: State) -> State:
+        """Computes if the game must be stopped because the agents crashed from a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def _compute_truncation(self, state: State) -> State:
+        """Computes if the game must be stopped because it is too long form a given state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def reset(self, key: jnp.ndarray) -> State:
+        """Resets the environment in initial state. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def auto_reset(self, state: State) -> State:
+        """Returns the State reinitialized if needed, else the actual State. Must be implemented in a subclass.
+
+        The values contained by State are passed in argument and used like a dictionary
+        because auto_reset is meant to be used by vmap and vmap doesn't accept objects.
+        """
+        raise NotImplementedError
+
+    def state_to_dict(self, state: State) -> dict:
+        """Translates the State into a dict. Must be implemented in a subclass."""
+        raise NotImplementedError
+
+    def step_vmap(self, action: jnp.ndarray, key: jnp.ndarray, **state_val) -> State:
+        """Used to vmap step, takes the values of the state and calls step with a new State object containing the state values. Must be implemented in a subclass.
```

Comment:
Add something like : "JAX's vmap cannot operate on array-of-structs, but can operate on struct-of-arrays, so the states actually contain array of arrays after vmap. Our solution to this is to convert the struct into a dictionary of array of array and plug it into the vmapped function as kwargs. This way, each value of the kwargs (the state members) will be processed as a regular array."

Reply:
Should we add that vmap can return a struct of arrays but can't take it as argument (it should be arrays) ?